### ISSUE TITLE
Validating request from another subdomain

### ISSUE THREAD

**Issue by JerryBels** [bshaffer/oauth2-server-php#596]
Hello, and thank you for this awesomeness !

I'm trying to implement an API where the users would authenticate using oauth2 tokens. The thing is, I want the API to be on its own subdomain (api.example.com) and the oauth2 server on another (auth.example.com). My API is built using Laravel, but on the auth subdomain I simply followed your step-by-step http://bshaffer.github.io/oauth2-server-php-docs/cookbook/

So, I successfully got a token back using Postman to test the URL auth.example.com/token.php with id, pass and grant_type in post. The question is : 

How, in my other subdomain api.example.com (or even on the main domain example.com, that will handle the website later, built also with Laravel) can I check the token and who it is authenticating ?

Thanks a lot ahead !

**Comment by JerryBels** [comment]
Bump !

Also, something more : in the step-by-step tutorial the client_secret is stored directly. I guess it's for demo purpose, and it actually needs to be encoded in sha1 or something, right ?

**Comment by phindmarsh** [comment]
Hey, on my phone so excuse the brevity.
You should look into the JWT access token stuff. It allows your auth server to issue signed tokens that you can later validate with only a JWT lib (such as fireball/jwt) and the signing key. 
I use this exact method for an API where the service issuing tokens is separate from the api that uses them for auth. All that I share between is an hmac secret used to sign the JWT tokens. If the signature verifies then you can trust the user_id etc supplied in the token.







On Tue, May 26, 2015 at 2:06 AM -0700, "JerryBels" <notifications@github.com> wrote:












Bump !



Also, something more : in the step-by-step tutorial the client_secret is stored directly. I guess it's for demo purpose, and it actually needs to be encoded in sha1 or something, right ?

**Comment by JerryBels** [comment]
Thanks for helping ! It looks a little complicated but I will try it. On the API side, in order to check the token, I have to implement the server in my Laravel like it's shown here http://bshaffer.github.io/oauth2-server-php-docs/cookbook/laravel/ ?

**Comment by JohnAdamsy** [comment]
Also adding onto @JerryBels request'..is there a way to set this up on a separate server, simply: the Authorization Server (AS) and then have several Resource Servers (RS), simply: the APIs on other servers, such that API requests from the client are handled by the single AS. And in mind, suppose you have multiple Resource Servers not necessarily implemented in PHP, say for instance, RS 1 is Node JS,  RS 2 is running Ruby on Rails, etc.

**Comment by phindmarsh** [comment]
@JohnAdamsy thats pretty much exactly what I outlined above. 

You have your AS issue [JWT access tokens](http://bshaffer.github.io/oauth2-server-php-docs/overview/jwt-access-tokens/), which are JSON objects that have been base 64 encoded and signed with a secret key. Then all you need is a JWT lib on your RS (in any language), decode the token, verify the signature, and if all good you can trust the contents of the token to be what the AS issued in the first place. If you have AS and RS that are physically separated and you do not wish to write a bridge to the AS for each request to the RS then this is the way to do it. As I said, I've implemented it and it works perfectly. 

@JerryBels On that note, if the codebase issuing tokens (authorization server) is the same as your API (resource server) then you can use the `verifyResourceRequest` method of the Resource Controller to do exactly that (see the _Create a Resource Controller_ of the [cookbook page](http://bshaffer.github.io/oauth2-server-php-docs/cookbook/)) . That only falls over if you have different codebases/apps, as hinted by @JohnAdamsy above). In that case I'd recommend the JWT token route. It is very simple to implement, just follow the [JWT access token docs](http://bshaffer.github.io/oauth2-server-php-docs/overview/jwt-access-tokens/).

**Comment by JohnAdamsy** [comment]
@phindmarsh Ah good stuff. Let me get onto it and will get back. Thanks a lot for the clarification.

**Comment by JerryBels** [comment]
@phindmarsh thank you for helping ! So in my case, where I have the same database but on different codebases, what would you recommend ?

Also, the questions I asked before : 
"On the API side, in order to check the token, I have to implement the server in my Laravel like it's shown here http://bshaffer.github.io/oauth2-server-php-docs/cookbook/laravel/ ?"

And : 
"in the step-by-step tutorial the client_secret is stored directly. I guess it's for demo purpose, and it actually needs to be encoded in sha1 or something, right ?"

And then I have a confusion : are the client_id and client_secret said to be the user's credentials ? I mean, the username and password he sends from the authentication form ?

**Comment by phindmarsh** [comment]
Well thats probably up to you, if I were in your position I'd probably just use the JWT tokens and pretend I don't share a database. 

I'm not that familiar with Laravel so I can't really help you with that, except to say that the docs are usually pretty accurate and that looks about right to me. It looks like you use that singleton method to configure your server, so any of the config code you see should probably go inside that block. 

On the client secret, yes you probably should pass that through the same type of security you use for passwords, the only caveat being if you need to display that client_secret via a UI (if you allow third party developers to create clients etc) If you do that and use standard password hashing (like bcrypt/scrypt) you won't be able to reverse it. In that case you can either use two-way encryption (AES or some such), but thats probably outside the scope of this discussion. If you are only ever going to have first-party clients (i.e. ones you or your team write) and you don't need to discover the client_secret once it has been generated, then sure, hash away.

On your last point, no, `client_id` and `client_secret` are used to identify clients only. By default a user's credentials are stored in the `oauth_users` table (if you are using MySQL for example). Typically this is probably overridden by most developers to use their own user store. When authenticating a user some grant types will require both client credentials and user credentials. Client credentials are to authenticate the client (i.e. the application code, mobile app etc) is authorized, and user credentials are used the the traditional sense to validate the user is who they say they are.

**Comment by JerryBels** [comment]
You helped me a lot, really. I will go back to work, thank you so much !

**Comment by JerryBels** [comment]
A little thing : is it possible to use JWT with user credentials grant ?

**Comment by bshaffer** [comment]
You can use Jwt AccessTokens with ANY grant type. But be careful not to
confuse this with JWT Bearer, which is its own specific kind of grant type

**Comment by JerryBels** [comment]
Thank you @bshaffer !

I made some progress. Now I successfully get a token using user credentials. But I would like to use another table, not oauth_users, to get the users and verify them.

I have read this : 

    Note: User storage is highly customized for each application, so it is highly recommended you implement your own storage using OAuth2\Storage\UserCredentialsInterface

But since I'm not sure what it means, I continued searching, to find that I can replace the table using an array of options like that : 

    $storage = new OAuth2\Storage\Pdo(array('dsn' => $dsn, 'username' => $username, 'password' => $password), array('user_table' => 'users'));

The issue is that in the storage/Pdo script it uses the fields from the oauth_users table, not from the one I just entered, of course. For example, it uses username, where in my users table I use email, that is the default of Laravel.

What would be the solution about that ?

Thanks again !
